### PR TITLE
Vid page

### PR DESCRIPTION
**Pull request by Yacine-Saoudi**
-added new videoSchema that references User
-added additional field for list of videos posted by user in the userSchema
-added middleware functions that decode any token sent from the frontend
-created new video controller that manages all post and get requests
-post requests create a new video
-get requests return an array of videos the user posted

Closes: #30, Closes: #29, Closes: #28, Closes: #27, Closes: #20

Signed-off-by: Yacine Saoudi yac.sao453@gmail.com

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -29,12 +30,15 @@ app.use(cors())
...
+app.use(middleware.tokenGet)
+app.use(middleware.userGet)
+
 /*
  * Routers
  */
 app.use('/login', loginRouter)
 app.use('/signup', signupRouter)
-
+app.use('/video', videoRouter)
```
Comment (OussamaSaoudi) [inline app.js:41]: Add a newline under video

File: controllers/login.js [hunk 2]
```diff
@@ -18,10 +18,9 @@ const User = require('../models/user')
 loginRouter.post('/', async (request, response) => {
   const body = request.body
   const user = await User.findOne({ username: body.username })
-
   const isValid = await bcrypt.compare(body.password, user.passHash)
   if (!(user && isValid)) {
-    return response.status(401).json({
+    return response.status(400).json({
```
Comment (OussamaSaoudi) [inline controllers/login.js:23]: How come you're changing to 400? 401 seems to make more sense here https://restfulapi.net/http-status-codes/

File: controllers/login.js [hunk 3]
```diff
@@ -37,4 +36,6 @@ loginRouter.post('/', async (request, response) => {
     .send ({ token })
 })
 
+
```
Comment (OussamaSaoudi) [inline controllers/login.js:39]: Remove the two newlines

File: controllers/video.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+/**
+ * @file Handles requests pertaining to user videos posting and retrieving all the videos.
+ */
+const videoRouter = require('express').Router()
+const User = require('../models/user')
+const Video = require('../models/video')
+/**
```
Comment (OussamaSaoudi) [inline controllers/video.js:7]: In general, you want chunks of (code + comments) separated on either side by a newline. So under `const Vido = require(..)` Add a newline

File: controllers/video.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+videoRouter.get('/', async (request, response) => {
+  const user = request.user
+
+  const videos = await Video.find({ user })
+
+  response
+    .status(200)
+    .json(videos)
+})
+/**
```
Comment (OussamaSaoudi) [inline controllers/video.js:23]: newline

File: controllers/video.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  // eslint-disable-next-line no-unused-vars
+  const newVid = new Video ({
+    username: request.user.username,
+    userId: request.user.id
+  })
+  await User
+    .findById(request.user.id)
+    .populate('videos')
+    .exec(User.vidoes.concat(newVid))
+    /* make it find users based on id */
```
Comment (OussamaSaoudi) [inline controllers/video.js:43]: Put this comment above the await User. Good comment tho

File: models/video.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+videoSchema.set('toJSON', {
+  transform: (document, returnedObject) => {
+    returnedObject.id = returnedObject._id.toString()
+    delete returnedObject._id
+    delete returnedObject.__v
+  }
+})
+
+const Video = mongoose.model('Video', videoSchema)
+
```
Comment (OussamaSaoudi) [inline models/video.js:31]: Remove one newline

File: package.json [hunk 8]
```diff
@@ -31,11 +31,14 @@
...
     "bcrypt": "^5.0.1",
     "cors": "^2.8.5",
+    "dom": "0.0.3",
     "dotenv": "^10.0.0",
     "express": "^4.17.1",
     "jsonwebtoken": "^8.5.1",
     "mongoose": "^5.12.13",
-    "mongoose-unique-validator": "^2.0.3"
+    "mongoose-unique-validator": "^2.0.3",
+    "react-router": "^5.2.0",
```
Comment (OussamaSaoudi) [inline package.json:40]: Remove react-router, react-router-dom, and dom. These are front end dependencies.

File: utils/middleware.js [hunk 9]
```diff
@@ -16,6 +17,41 @@ const reqLog = (request, response, next) => {
   console.log('Method: ', request.method)
   console.log('Path: ', request.path)
   console.log('Body: ', request.body)
+  console.log('test')
```
Comment (OussamaSaoudi) [inline utils/middleware.js:20]: Remove `console.log('test')`

File: utils/middleware.js [hunk 10]
```diff
@@ -16,6 +17,41 @@ const reqLog = (request, response, next) => {
...
+  next()
+}
```
Comment (OussamaSaoudi) [inline utils/middleware.js:22]: newline

File: utils/middleware.js [hunk 11]
```diff
@@ -16,6 +17,41 @@ const reqLog = (request, response, next) => {
...
+ *
+ * @param {object} request Receives the token.
+ * @param {object} response Returns the token as request.token.
+ * @param {Function} next Next middleware fuunction to be called.
+ */
+const tokenGet = (request, response, next) => {
+  const auth = request.get('Authorization')
+  if (auth){
+    request.token = auth.substring(7)
+
```
Comment (OussamaSaoudi) [inline utils/middleware.js:34]: remove the newline

File: utils/middleware.js [hunk 12]
```diff
@@ -16,6 +17,41 @@ const reqLog = (request, response, next) => {
...
+/**
+ * Token receiver that takes the request, finds the auth header, and recieves the token while isolating the bearer token.
+ *
+ * @param {object} request Receives the token.
+ * @param {object} response Returns the token as request.token.
+ * @param {Function} next Next middleware fuunction to be called.
+ */
+const tokenGet = (request, response, next) => {
+  const auth = request.get('Authorization')
+  if (auth){
```
Comment (OussamaSaoudi) [inline utils/middleware.js:32]: add space after (auth)
`if (auth) {`

File: utils/middleware.js [hunk 13]
```diff
@@ -16,6 +17,41 @@ const reqLog = (request, response, next) => {
...
+    request.token = auth.substring(7)
+
+  }
+  next()
+}
```
Comment (OussamaSaoudi) [inline utils/middleware.js:37]: newline

File: utils/middleware.js [hunk 14]
```diff
@@ -16,6 +17,41 @@ const reqLog = (request, response, next) => {
...
+/**
+ * Token decoder that takes the token, verifies it, then decodes it and saves it as a user object and returns that.
+ *
+ * @param {object} request Receives the token minus the bearer substring.
+ * @param {object} response Returns request.user object that contains the user's username and id.
+ * @param {Function} next Next middleware fuunction to be called.
+ */
+const userGet = (request, response, next) => {
+  if (request.token){
```
Comment (OussamaSaoudi) [inline utils/middleware.js:46]: Space after (request.token)
`if (request.token) {`

File: utils/middleware.js [hunk 15]
```diff
@@ -45,15 +81,22 @@ const errorHandler = (error, request, response, next) => {
   if (error.name === 'error') {
     return response.status(400).send({ error: 'error name' })
   }
-  else if (error.name == 'validationError'){
-    return response.status(401).send({ error: 'invalid username or password' })
+  else if (error.name == 'ValidationError'){
+    return response.status(400).send({ error: 'invalid username or password' })
```
Comment (OussamaSaoudi) [inline utils/middleware.js:85]: I think 401 might work better here. Comment here what case the JsonWebTokenError occurs.

File: utils/middleware.js [hunk 16]
```diff
@@ -45,15 +81,22 @@ const errorHandler = (error, request, response, next) => {
...
+  }
+  else if (error.name === 'JsonWebTokenError'){
```
Comment (OussamaSaoudi) [inline utils/middleware.js:87]: Space after 'JsonWebTokenError')

File: controllers/login.js (hunk 2 above)
Comment (Yacine-Saoudi) [reply controllers/login.js:23]: i dont think so because its an authorization error

File: package.json (hunk 8 above)
Comment (Yacine-Saoudi) [reply package.json:40]: yeah that was on accident as i was in the wrong terminal

**Pull request merged**